lib/auth/auth_provider.js

```diff
@@ -0,0 +1,177 @@
+'use strict';
+
+const MongoError = require('../error').MongoError;
+const Query = require('../connection/commands').Query;
+
+/**
+ * Creates a new AuthProvider, which dictates how to authenticate for a given
+ * mechanism.
+ * @class
+ */
+class AuthProvider {
+  constructor(bson) {
+    this.bson = bson;
+    this.authStore = [];
+  }
+
+  _buildCommand(namespace, cmd) {
```

Comment:
lacks docs, I do expect that this abstracts creating a Query so that's cool (for e.g. OP_MSG in the future)